### DISCUSSION TITLE
Allow custom React components, not just full Apps

### DISCUSSION THREAD

**Discussion by he-la** [windmill-labs/windmill#2191] [Questions] [answered]
My understanding is that windmill currently allows importing full React apps, which internally can invoke flows etc on windmill.

It would be nice if we could use custom React components also in the App builder, passing props to the component based on the inputs and have some way to declare the component's outputs.

Our use case for this would be (among others) a custom rich text editor that we need to deal with our rich text format. Most of our internal tools are basic CRUD, so a combined API+App builder like windmill would be stellar - but the need for custom components for some rare but important cases is a bit of a showstopper to going all-in.

AFAIK Retool supports custom React components, which makes it much easier to commit to since we can know that we'll be able to unblock if we run into specific needs like this. But of course it lacks everything else we'd get from the windmill platform.

**Comment by rubenfiszel** [comment, marked as answer]
Hi @he-la ,

We will support this and are rolling it out this week to our clients but it is an enterprise feature: https://github.com/windmill-labs/windmill/pull/2173

**Reply by he-la** [reply]
Gotcha, thanks! I guess that makes sense to have it as an EE feature!
